tests/std/tests/P0896R4_ranges_subrange/test.cpp

```diff
@@ -0,0 +1,1223 @@
+            STATIC_ASSERT(same_as<T, subrange<I, S, subrange_kind::sized>>);
+
+            STATIC_ASSERT(range<T>);
+            STATIC_ASSERT(!copyable<I> || range<T const&>);
+
+            if constexpr (borrowed_range<Rng>) {
+                using U = decltype(subrange(std::declval<Rng>(), std::declval<size_type>()));
+                STATIC_ASSERT(same_as<U, T>);
+            }
+        }
+
+        return true;
+    }
+    STATIC_ASSERT(test_ctad<test_range<output_iterator_tag, int, false, false>>());
+    STATIC_ASSERT(test_ctad<test_range<output_iterator_tag, int, true, false>>());
+    STATIC_ASSERT(test_ctad<test_range<input_iterator_tag, int, false, false>>());
+    STATIC_ASSERT(test_ctad<test_range<input_iterator_tag, int, true, false>>());
+    STATIC_ASSERT(test_ctad<test_range<forward_iterator_tag, int, false, false>>());
+    STATIC_ASSERT(test_ctad<test_range<forward_iterator_tag, int, false, true>>());
+    STATIC_ASSERT(test_ctad<test_range<forward_iterator_tag, int, true, false>>());
+    STATIC_ASSERT(test_ctad<test_range<forward_iterator_tag, int, true, true>>());
+    STATIC_ASSERT(test_ctad<test_range<bidirectional_iterator_tag, int, false, false>>());
+    STATIC_ASSERT(test_ctad<test_range<bidirectional_iterator_tag, int, false, true>>());
+    STATIC_ASSERT(test_ctad<test_range<bidirectional_iterator_tag, int, true, false>>());
+    STATIC_ASSERT(test_ctad<test_range<bidirectional_iterator_tag, int, true, true>>());
+    STATIC_ASSERT(test_ctad<test_range<random_access_iterator_tag, int, true, false>>());
+    STATIC_ASSERT(test_ctad<test_range<random_access_iterator_tag, int, true, true>>());
+    STATIC_ASSERT(test_ctad<test_range<contiguous_iterator_tag, int, true, false>>());
+    STATIC_ASSERT(test_ctad<test_range<contiguous_iterator_tag, int, true, true>>());
+
+    STATIC_ASSERT(test_ctad<std::forward_list<int>>());
+    STATIC_ASSERT(test_ctad<std::list<int>>());
+    STATIC_ASSERT(test_ctad<std::vector<int>>());
+    STATIC_ASSERT(test_ctad<std::string_view>());
+
+    // Validate conversion to PairLike
+    STATIC_ASSERT(std::is_convertible_v<subrange<int*>, std::pair<int*, int*>>);
+    STATIC_ASSERT(!std::is_convertible_v<subrange<int const*>, std::pair<int*, int*>>);
+    STATIC_ASSERT(std::is_convertible_v<subrange<int*>, std::pair<int const*, int const*>>);
+    STATIC_ASSERT(std::is_convertible_v<subrange<int const*>, std::pair<int const*, int const*>>);
+    STATIC_ASSERT(std::is_convertible_v<subrange<int*>, std::tuple<int*, int*>>);
+    STATIC_ASSERT(!std::is_convertible_v<subrange<int const*>, std::tuple<int*, int*>>);
+    STATIC_ASSERT(std::is_convertible_v<subrange<int*>, std::tuple<int const*, int const*>>);
+    STATIC_ASSERT(std::is_convertible_v<subrange<int const*>, std::tuple<int const*, int const*>>);
+
+    constexpr bool test_advance_next_prev() {
+        // Validate advance/next/prev
+        int const some_ints[] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};

```

Comment:
Doesn't really matter for test code, but `static constexpr` (or at least `static const`) is more efficient here.

Reply:
`constexpr` functions may not have local variables with `static` storage duration per [dcl.constexpr]/3.5.3. I'd make a named global for this in product code, but for test code I prefer the encapsulation.